src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -198,22 +186,68 @@ public void visitToken(DetailAST ast) {
         }
     }
 
+    /**
+     * Called to process a type definition.
+     *
+     * @param ast the token to process
+     */
+    private void visitClass(DetailAST ast) {
+        final String qualifiedClassName = getQualifiedClassName(ast);
+        final ClassDesc currClass = new ClassDesc(qualifiedClassName, classes.size(), ast);
+        classes.push(currClass);
+        innerClasses.put(qualifiedClassName, currClass);
+    }
+
+    /**
+     * Called to process a constructor definition.
+     *
+     * @param ast the token to process
+     */
+    private void visitCtor(DetailAST ast) {
+        if (!ScopeUtil.isInEnumBlock(ast) && !ScopeUtil.isInRecordBlock(ast)) {
+            final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
+            final ClassDesc desc = classes.peek();
+            if (modifiers.findFirstToken(TokenTypes.LITERAL_PRIVATE) == null) {
+                desc.registerNonPrivateCtor();
+            }
+            else {
+                desc.registerPrivateCtor();
+            }
+        }
+    }
+
     @Override
     public void leaveToken(DetailAST ast) {
         if (ast.getType() == TokenTypes.CLASS_DEF) {
-            final ClassDesc desc = classes.pop();
-            if (desc.isWithPrivateCtor()
+            classes.pop();
+        }
+        if (TokenUtil.isRootNode(ast.getParent())) {
+            innerClasses.forEach((key, value) -> {
+                final DetailAST classAst = value.getClassAst();
+                registerNestedSubclassToOuterSuperClasses(classAst, key);
+            });
+            innerClasses.forEach((key, value) -> {
```

Comment:
Just checking - we cannot combine bodies of lambdas because all subclasses must be registered before logging, right?

Reply:
Yes. The first pass to mark some classes as a "super class for some other class", the second pass to report all classes that should be final.

Reply:
Resolved